klippa/src/name.rs

```diff
@@ -0,0 +1,123 @@
+//! impl subset() for name table
+
+use std::cmp::Ordering;
+
+use crate::{
+    Plan,
+    SubsetError::{self, SubsetTableError},
+    SubsetFlags,
+};
+
+use write_fonts::{
+    read::{tables::name::Name, FontRef, TableProvider, TopLevelTable},
+    FontBuilder,
+};
+
+// reference: subset() for name table in harfbuzz
+// https://github.com/harfbuzz/harfbuzz/blob/main/src/OT/name/name.hh#L387
+pub(crate) fn subset_name(
+    font: &FontRef,
+    plan: &Plan,
+    builder: &mut FontBuilder,
+) -> Result<(), SubsetError> {
+    let name = font.name().or(Err(SubsetTableError(Name::TAG)))?;
+    let name_records = name.name_record();
+    //TODO: support name_table_override
+    //TODO: support name table version 1
+    let mut retained_name_record_idxes = name_records
+        .iter()
+        .enumerate()
+        .filter(|(_, record)| {
+            if !plan.name_ids.contains(record.name_id())
+                || !plan.name_languages.contains(record.language_id())
+            {
+                return false;
+            }
+            plan.subset_flags
+                .contains(SubsetFlags::SUBSET_FLAGS_NAME_LEGACY)
+                || record.is_unicode()
+        })
+        .map(|x| x.0)
+        .collect::<Vec<_>>();
+
+    retained_name_record_idxes.sort_unstable_by(|a, b| {
+        let id_a = name_records[*a].platform_id();
+        let id_b = name_records[*b].platform_id();
+        if id_a != id_b {
+            return id_a.cmp(&id_b);
+        }
+
+        let id_a = name_records[*a].encoding_id();
+        let id_b = name_records[*b].encoding_id();
+        if id_a != id_b {
+            return id_a.cmp(&id_b);
+        }
+
+        let id_a = name_records[*a].language_id();
+        let id_b = name_records[*b].language_id();
+        if id_a != id_b {
+            return id_a.cmp(&id_b);
+        }
+
+        let id_a = name_records[*a].name_id().to_u16();
+        let id_b = name_records[*b].name_id().to_u16();
+        if id_a != id_b {
+            return id_a.cmp(&id_b);
+        }
+
+        let len_a = name_records[*a].length();
+        let len_b = name_records[*b].length();
+        if len_a != len_b {
+            return len_a.cmp(&len_b);
+        }
+
+        Ordering::Equal
+    });
```

Comment:
This can be replaced by

```rust
    retained_name_record_idxes.sort_unstable_by_key(|nr| {
        let nr = name_records[*nr];
        (nr.platform_id(), nr.encoding_id(), nr.language_id(), nr.name_id().to_u16(), nr.length())
    });
```